File: cpp/054_SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ret;
        if(matrix.empty()) return ret;
        int right = matrix[0].size()-1, left = 0;
        int up = 0, down = matrix.size()-1;
        while(1){
            //to right
            if(left > right) return ret;
            for(int i = left; i <= right; ++i){
                ret.push_back(matrix[up][i]);
            }
            up++;
            //to down
            if(up > down) return ret;
            for(int i = up; i <= down; ++i){
                ret.push_back(matrix[i][right]);
            }
            right--;
            //to left
            if(left > right) return ret;
            for(int i = right; i >= left; --i){
                ret.push_back(matrix[down][i]);
            }
            down--;
            //to up
            if(up > down) return ret;
            for(int i = down; i >= up; --i){
                ret.push_back(matrix[i][left]);
            }
            left++;
        }
        return ret;
    }
};

File: js/054_SpiralMatrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    var ret = [];
    if(matrix.length < 1) return ret;
    var left = 0; right = matrix[0].length-1;
    var up = 0, down = matrix.length-1;
    while(true){
        //to right
        if(left > right) break;
        for(var r = left; r <= right; ++r){
            ret.push(matrix[up][r]);
        }
        up++;
        //to down
        if(up > down) break;
        for(var d = up; d <= down; ++d){
            ret.push(matrix[d][right]);
        }
        right--;
        //to left
        if(left > right) break;
        for(var l = right; l >= left; --l){
            ret.push(matrix[down][l]);
        }
        down--;
        //to up
        if(up > down) break;
        for(var u = down; u >= up; --u){
            ret.push(matrix[u][left]);
        }
        left++;
    }
    return ret;
};